source/common/ext_authz/ext_authz_impl.cc

```diff
@@ -0,0 +1,241 @@
+
+#include "common/common/assert.h"
+#include "common/grpc/async_client_impl.h"
+#include "common/http/headers.h"
+#include "common/protobuf/protobuf.h"
+
+#include "fmt/format.h"
+
+namespace Envoy {
+namespace ExtAuthz {
+
+using ::envoy::api::v2::auth::AttributeContext;
+using ::envoy::api::v2::auth::AttributeContext_HttpRequest;
+using ::envoy::api::v2::auth::AttributeContext_Peer;
+using ::envoy::api::v2::auth::AttributeContext_Request;
+
+GrpcClientImpl::GrpcClientImpl(Grpc::AsyncClientPtr&& async_client,
+                               const Optional<std::chrono::milliseconds>& timeout)
+    : service_method_(*Protobuf::DescriptorPool::generated_pool()->FindMethodByName(
+          "envoy.api.v2.auth.Authorization.Check")),
+      async_client_(std::move(async_client)), timeout_(timeout) {}
+
+GrpcClientImpl::~GrpcClientImpl() { ASSERT(!callbacks_); }
+
+void GrpcClientImpl::cancel() {
+  ASSERT(callbacks_ != nullptr);
+  request_->cancel();
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::check(RequestCallbacks& callbacks,
+                           const envoy::api::v2::auth::CheckRequest& request,
+                           Tracing::Span& parent_span) {
+  ASSERT(callbacks_ == nullptr);
+  callbacks_ = &callbacks;
+
+  request_ = async_client_->send(service_method_, request, *this, parent_span, timeout_);
+}
+
+void GrpcClientImpl::onSuccess(std::unique_ptr<envoy::api::v2::auth::CheckResponse>&& response,
+                               Tracing::Span& span) {
+  CheckStatus status = CheckStatus::OK;
+  ASSERT(response->status().code() != Grpc::Status::GrpcStatus::Unknown);
+  if (response->status().code() != Grpc::Status::GrpcStatus::Ok) {
+    status = CheckStatus::Denied;
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceUnauthz);
+  } else {
+    span.setTag(Constants::get().TraceStatus, Constants::get().TraceOk);
+  }
+
+  callbacks_->complete(status);
+  callbacks_ = nullptr;
+}
+
+void GrpcClientImpl::onFailure(Grpc::Status::GrpcStatus status, const std::string&,
+                               Tracing::Span&) {
+  ASSERT(status != Grpc::Status::GrpcStatus::Ok);
+  UNREFERENCED_PARAMETER(status);
+  callbacks_->complete(CheckStatus::Error);
+  callbacks_ = nullptr;
+}
+
+GrpcFactoryImpl::GrpcFactoryImpl(const envoy::api::v2::GrpcService& grpc_service,
+                                 Grpc::AsyncClientManager& async_client_manager,
+                                 Stats::Scope& scope) {
+  async_client_factory_ = async_client_manager.factoryForGrpcService(grpc_service, scope);
+}
+
+ClientPtr GrpcFactoryImpl::create(const Optional<std::chrono::milliseconds>& timeout) {
+  return std::make_unique<GrpcClientImpl>(async_client_factory_->create(), timeout);
+}
+
+std::unique_ptr<::envoy::api::v2::Address> ExtAuthzCheckRequestGenerator::getProtobufAddress(
+    const Network::Address::InstanceConstSharedPtr& instance) {
+
+  auto addr = std::make_unique<::envoy::api::v2::Address>();
```

Comment:
I think it's more natural in protobuf to invert this construction. What you have is: 
1. getConnectionPeer calls getProtobufAddress to get some remote address PB.
2. getConnectionPeer then does a set_allocated_address with the result.

A much more common pattern is.
1. getConnectionPeer calls `mutable_allocated_addresss()` to allocate memory and obtain a pointer.
2. getProtobufAddress takes this pointer and then writes directly into the memory.

Of course, you would probably rename the methods. See https://github.com/envoyproxy/envoy/blob/master/source/common/config/address_json.cc#L10 and its callers for an example.

What you have is correct, I just find it noisier, since we have to deal with smart pointer creation and release, rather than idiomatically having the protobuf library `mutable_foo()` allocators do their work.